src/Common/ConnectionPool.cpp

```diff
@@ -0,0 +1,1075 @@
+        {
+            case DB::ConnectionGroupType::STORAGE:
+                return getMetricsForS3StorageConnectionPool();
+            case DB::ConnectionGroupType::DISK:
+                return getMetricsForS3DiskConnectionPool();
+            case DB::ConnectionGroupType::HTTP:
+                return getMetricsForHTTPConnectionPool();
+        }
+    }
+
+
+    class ConnectionGroup
+    {
+    public:
+        using Ptr = std::shared_ptr<ConnectionGroup>;
+
+        explicit ConnectionGroup(DB::ConnectionGroupType type_)
+            : type(type_)
+            , metrics(getConnectionPoolMetrics(type_))
+        {}
+
+        void setLimits(DB::ConnectionPoolLimits limits_)
+        {
+            std::lock_guard lock(mutex);
+            limits = std::move(limits_);
+            mute_warning_until = 0;
+        }
+
+        bool isSoftLimitReached() const
+        {
+            std::lock_guard lock(mutex);
+            return active_connections_in_group >= limits.soft_limit;
+        }
+
+        bool isHardLimitReached() const
+        {
+            std::lock_guard lock(mutex);
+            return active_connections_in_group >= limits.hard_limit;
+        }
+
+        void atConnectionCreate()
+        {
+            std::lock_guard lock(mutex);
+
+            ++active_connections_in_group;
+
+            if (active_connections_in_group >= limits.warning_limit && active_connections_in_group >= mute_warning_until)
+            {
+                LOG_WARNING(log, "Too many active sessions in group {}, count {}", type, active_connections_in_group);
+                mute_warning_until = roundUp(active_connections_in_group, 100);
+            }
+        }
+
+        void atConnectionDestroy()
+        {
+            std::lock_guard lock(mutex);
+
+            --active_connections_in_group;
+
+            if (active_connections_in_group < limits.warning_limit && mute_warning_until > 0)
+            {
+                LOG_WARNING(log, "Sessions count is OK in the group {}, count {}", type, active_connections_in_group);
+                mute_warning_until = 0;
+            }
+        }
+
+        void addConnections(size_t num)
+        {
+            std::lock_guard lock(mutex);
+            active_connections_in_group += num;
+        }
+
+        void subConnections(size_t num)
+        {
+            std::lock_guard lock(mutex);
+            if (num > active_connections_in_group)
+                active_connections_in_group -= num;
+            else
+                active_connections_in_group = 0;
+        }
+
+        DB::ConnectionGroupType getType() const
+        {
+            return type;
+        }
+
+        const DB::ConnectionPoolMetrics & getMetrics() const
+        {
+            return metrics;
+        }
+
+    private:
+
+        const DB::ConnectionGroupType type;
+        const DB::ConnectionPoolMetrics metrics;
+
+        LoggerPtr log = ::getLogger("ConnectionGroup");
+
+        mutable std::mutex mutex;
```

Comment:
Would be nice to say something somewhere about the order of the mutexes (for avoiding deadlocks). It's ConnectionPools -> EndpointConnectionPool -> ConnectionGroup, right? E.g. (feel free to word better):
```suggestion
        // If multiple mutexes are held simultaneously, they should be locked in this order: ConnectionPools::mutex, then EndpointConnectionPool::mutex, then ConnectionGroup::mutex.
        mutable std::mutex mutex;
````